lib/debugger/debugger.dart

```diff
@@ -0,0 +1,1309 @@
+    _executionPointElement = null;
+
+    _refreshMarkers();
+  }
+}
+
+typedef URIDescriber = String Function(String uri);
+
+class BreakpointsView {
+  BreakpointsView(this._breakpointsCountDiv, DebuggerState debuggerState,
+      URIDescriber uriDescriber) {
+    _items = new SelectableList<Breakpoint>()
+      ..flex()
+      ..clazz('menu-item-bottom-border')
+      ..element.style.overflowY = 'scroll';
+
+    _items.setRenderer((Breakpoint breakpoint) {
+      final dynamic location = breakpoint.location;
+
+      final CoreElement element = li(text: '', c: 'list-item');
+
+      if (location is UnresolvedSourceLocation) {
+        element.text = uriDescriber(location.script.uri);
+        element.add(span(text: ' line ${location.line}', c: 'subtle'));
+      } else if (location is SourceLocation) {
+        element.text = uriDescriber(location.script.uri);
+
+        // Modify the rendering slightly asynchronously.
+        debuggerState.getScript(location.script).then((Script script) {
+          final Pos pos =
+              debuggerState.calculatePosition(script, location.tokenPos);
+          element.add(span(text: ' line ${pos.line}', c: 'subtle'));
+        });
+      }
+
+      if (!breakpoint.resolved) {
+        element.add(span(text: ' (unresolved)', c: 'subtle'));
+      }
+
+      return element;
+    });
+  }
+
+  Stream<Breakpoint> get onDoubleClick => _items.onDoubleClick;
+
+  final CoreElement _breakpointsCountDiv;
+
+  SelectableList<Breakpoint> _items;
+
+  CoreElement get element => _items;
+
+  Stream<Breakpoint> get onSelectionChanged => _items.onSelectionChanged;
+
+  void showBreakpoints(List<Breakpoint> breakpoints) {
+    breakpoints = breakpoints.toList();
+    breakpoints.sort(_breakpointComparator);
+
+    _items.setItems(breakpoints);
+    _breakpointsCountDiv.text = breakpoints.length.toString();
+  }
+}
+
+class ScriptsView {
+  ScriptsView(URIDescriber uriDescriber) {
+    _items = new SelectableList<ScriptRef>()
+      ..flex()
+      ..element.style.overflowY = 'scroll';
+    _items.setRenderer((ScriptRef scriptRef) {
+      final String uri = scriptRef.uri;
+      final String name = uriDescriber(uri);
+      final CoreElement element = li(text: name, c: 'list-item');
+      if (name != uri) {
+        element.add(span(text: ' $uri', c: 'subtle'));
+      }
+      element.tooltip = uri;
+      return element;
+    });
+  }
+
+  SelectableList<ScriptRef> _items;
+
+  String rootLib;
+
+  List<ScriptRef> get items => _items.items;
+
+  CoreElement get element => _items;
+
+  Stream<ScriptRef> get onSelectionChanged => _items.onSelectionChanged;
+
+  void showScripts(
+    List<ScriptRef> scripts,
+    String rootLib,
+    String commonPrefix, {
+    bool selectRootScript = false,
+  }) {
+    this.rootLib = rootLib;
+
+    scripts.sort((ScriptRef ref1, ScriptRef ref2) {
+      String uri1 = ref1.uri;
+      String uri2 = ref2.uri;
+
+      if (uri1.startsWith('dart:_')) {
```

Comment:
add a helper method to cleanup dart:_ uris

Reply:
done